Fix cn bugs

tacaswell opened the pull request:
Fixes two bugs:

 - as was noted by @anntzer that the caching mechanism in `to_rgb` meant that the CN mapping is only done once
 - tweak validation in rcsetup for colors that go into `prop_cycle` to forbid setting `CN` style values (due to circularness)

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -29,11 +29,12 @@
     import collections as abc
 from matplotlib.fontconfig_pattern import parse_fontconfig_pattern
 from matplotlib.colors import is_color_like
+import re
```
`re` should be above `matplotlib` imports.

WeatherGod replied:
This import is still in the wrong location

tacaswell replied:
I got the other one

QuLogic commented on lib/matplotlib/tests/test_rcparams.py:
```diff
@@ -417,6 +418,5 @@ def test_rcparams_reset_after_fail():
 
         assert mpl.rcParams['text.usetex'] is False
 
-
```
Should not remove this blank line.

WeatherGod commented on lib/matplotlib/colors.py:
```diff
@@ -194,14 +194,17 @@ def to_rgb(self, arg):
                     try:
                         argl = self._parse_nth_color(arg)
+                        # in this case we do not want to cache in case
+                        # the rcparam changes, recurse with the actually color
```
"actually" -> "actual"

WeatherGod commented on lib/matplotlib/rcsetup.py:
```diff
@@ -818,6 +836,12 @@ def validate_cycler(s):
+        _prop_validators[key](vals)
+
+    if 'color' not in cycler_inst.keys:
+        cycler_inst = cycler_inst * cycler('color', ['k'])
```
I honestly do not agree with this at all. This will cause unexpected behavior, which we noted already because the cycler will advance when the user did not expect it to. Why can't CN just fall back to the default linecolor (which is always defined) if the property cycle doesn't have color?

What I would agree with is if the cycler has *nothing*, then it should fall back to something. What that something is, I am not sure (hard to reference an rcParam while in the middle of processing them...).

tacaswell replied:
This is what is currently in the code.  I suggest we merge this (to un break tests) and then discuss this in a separate issue.

WeatherGod commented on lib/matplotlib/tests/test_cycles.py:
```diff
@@ -129,7 +129,7 @@ def test_property_collision_plot():
     ax.set_prop_cycle('linewidth', [2, 4])
     for c in range(1, 4):
-        ax.plot(np.arange(10), c * np.arange(10), lw=0.1)
+        ax.plot(np.arange(10), c * np.arange(10), lw=0.1, color='k')
```
This masks the problem exposed by the implicit inclusion of "color" into the property cycle.

tacaswell replied:
This means the test will pass independent of the decision on the other issue

WeatherGod replied:
This unit test is explicitly testing that the property cycler does not advance if the user specifies the property provided by the cycler. That was the original intent. The fact that it came about because of a slightly different bug doesn't matter.

tacaswell replied:
This still tests that, just have to include all of the keys in the cycler
and as implemented on v2.x and master, 'color' is always in the cycler

On Fri, May 6, 2016, 16:40 Benjamin Root <notifications@github.com> wrote:

> In lib/matplotlib/tests/test_cycles.py
> <https://github.com/matplotlib/matplotlib/pull/6374#discussion_r62388023>:
>
> > @@ -129,7 +129,7 @@ def test_property_collision_plot():
> >      fig, ax = plt.subplots()
> >      ax.set_prop_cycle('linewidth', [2, 4])
> >      for c in range(1, 4):
> > -        ax.plot(np.arange(10), c * np.arange(10), lw=0.1)
> > +        ax.plot(np.arange(10), c * np.arange(10), lw=0.1, color='k')
>
> This unit test is explicitly testing that the property cycler does not
> advance if the user specifies the property provided by the cycler. That was
> the original intent. The fact that it came about because of a slightly
> different bug doesn't matter.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/6374/files/faae5069bb987fd5447652860a8d1fb602c3289d#r62388023>
>

WeatherGod replied:
I have filed issue #6380 to track that. I consider it a blocker.

tacaswell commented:
fixed typos/import ordering and removed the commit that add the color cycler in `rcsetup` (as that breaks tests).

I think this should go in (despite @WeatherGod 's protests) to fix the test and we open a new issue to discuss how the default color is handled.

jenshnielsen commented:
Looks like this still have issues on python 2.7

WeatherGod commented:
yeah, those 2.7 failures are non-trivial. Something is seriously wrong.

WeatherGod commented:
I'll build and test this pr for v2.7 to see if I can figure it out

WeatherGod commented:
I can reproduce the two hatch validation errors, and one of the two image comparison failures. The plot for the linestyle cycle test is coming up blank... no lines, no legend. I can't reproduce the markerstyle test failure, though. Makes me wonder if something non-deterministic is happening?

WeatherGod commented:
I got other obligations for today, so I don't have more time to pursue this until tonight.

tacaswell commented:
The python2 issue was just using the wrong string type.

It makes me happy to find string/unicode issues that only show up on LPy :smile: .

efiring commented:
Given that this is passing tests now, I am going to go ahead and merge it so that testing of other PRs will no longer be broken.  Work can proceed towards solving #6380.

efiring commented:
I merged this into master.

WeatherGod commented:
I am still a little concerned about those non-deterministic errors. Could it be a problem with the caching mechanism?

efiring commented:
@weathergod, I thought the remaining Travis error (frequent but intermittent) was just the doc build error we have been seeing recently, which is a failure of Travis to download basemap.  However, the attempt by @jenshnielsen to fix that with #6379 failed Appveyor last night on 2.7, apparently because it didn't find latex.

WeatherGod commented:
Nope. As I noted earlier:

> and one of the two image comparison failures. The plot for the linestyle cycle test is coming up blank... no lines, no legend. I can't reproduce the markerstyle test failure, though. Makes me wonder if something non-deterministic is happening?

Now, the damnest thing is that I just tried out the branch with the last fix in there (which should have nothing to do with the remaining failure), and it passes -- multiple times. I don't get it.

tacaswell commented:
I think the problem is that there is a leak in decorators exception
handling. Because we are running the tests in parallel exactly which test
gets broken is a bit random.

On Sun, May 8, 2016, 13:39 Benjamin Root <notifications@github.com> wrote:

> Nope. As I noted earlier:
>
> and one of the two image comparison failures. The plot for the linestyle
> cycle test is coming up blank... no lines, no legend. I can't reproduce the
> markerstyle test failure, though. Makes me wonder if something
> non-deterministic is happening?
>
> Now, the damnest thing is that I just tried out the branch with the last
> fix in there (which should have nothing to do with the remaining failure),
> and it passes -- multiple times. I don't get it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/6374#issuecomment-217735518>
>

WeatherGod commented:
perhaps, but it isn't limited to parallel execution. Yesterday, I ran only `test_cycles.py` directly and got that one failure in addition to the two hatch property validation errors. Nothing was parallel. There may be a leak, but it is likely within the same process.
